### ISSUE TITLE
Q: Insert partially NULL binary string

### ISSUE THREAD

**Issue by mloskot** [nanodbc/nanodbc#122]
_From @surgura on September 29, 2016 7:56_

It can be useful to insert binary data that is partially NULL. E.g. 0x12 00 13 00 00 13 14 16
I am unsure if this feature is present. As far as I can tell it is only possible to insert either an all NULL value, or insert a c-like string(where NULL is the string terminator).

It appears to me that I must have missed something here?

_Copied from original issue: lexicalunit/nanodbc#209_

**Comment by mloskot** [comment]
On 29 Sep 2016 08:56, "Aart Stuurman" notifications@github.com wrote:

> It can be useful to insert binary data that is partially NULL. E.g. 0x12 00 13 00 00 13 14 16

I think you need to provide an example of SQL, details of table/column
schema to illustrate what you aim to achieve. Perhaps you have sample of
C++ code to show how you call nanodbc and what is not working.

**Comment by mloskot** [comment]
_From @surgura on September 29, 2016 11:47_

Sorry for being unclear. Lets see if I can make it more clear.

The C++ code below describes what I'm trying to achieve. The database column 'raw_data' is a Mysql BINARY(300) field(just for testing purposes).

The problem I have is that I cannot find out how to insert this binary data. I cannot convert this to a c-string because that is terminated by a zero byte.

```
// create my binary string
std::vector<unsigned char> rawdata;
rawdata.push_back(14);
rawdata.push_back(0);
rawdata.push_back(99);
rawdata.push_back(0);
rawdata.push_back(0);
rawdata.push_back(3);

// display string
for (unsigned char& c : rawdata)
    std::cout << (unsigned int)c << " ";
std::cout << std::endl;

// insert everything in database
nanodbc::statement statement(nanodbcInstance);
nanodbc::prepare(statement, "INSERT INTO mydb.mytable (raw_data, unpack_successful) VALUES (?, FALSE)");
// Bind raw data. (this does not compile because I do not use a supported type)
statement.bind(0, rawdata);
nanodbc::just_execute(statement);
```

**Comment by mloskot** [comment]
_From @surgura on October 3, 2016 12:44_

I figured out a way to bind this type of data. statement.bind can also be called by providing an array and the number of objects in the array.

```
// create my binary string
std::vector<unsigned char> rawdata;
rawdata.push_back(14);
rawdata.push_back(0);
rawdata.push_back(99);
rawdata.push_back(0);
rawdata.push_back(0);
rawdata.push_back(3);

// display string
for (unsigned char& c : rawdata)
    std::cout << (unsigned int)c << " ";
std::cout << std::endl;

// insert everything in database
nanodbc::statement statement(nanodbcInstance);
nanodbc::prepare(statement, "INSERT INTO mydb.mytable (raw_data, unpack_successful) VALUES (?, FALSE)");

vvvvv
// Bind raw data. (this does not compile because I do not use a supported type)
statement.bind(0, rawdata.data(), rawdata.size());
^^^^^
nanodbc::just_execute(statement);
```

This does however make for some strange results in the database for the parts of data that you don't write to. For example if BINARY(30) is used and you only bind an array of 10 long, the last 20 bytes seem to be gibberish. This seems logical to me, but when I used a TEXT field, I expected 10 bytes to be used in the database, where I instead got my 10 bytes, plus 30-ish extra bytes of gibberish, which I did not expect or add to my bound array.

Comments would be very much appreciated. Meanwhile I am experimenting and trying to figure out what I am doing wrong.

**Comment by mloskot** [comment]
_From @surgura on March 29, 2017 11:55_

Resolved by #219 solution.
